File: v1/Cpp/boj/Intermediate/boj_11437.cpp
//
// Created by seungsu on 12/27/22.
//

#include <bits/stdc++.h>
using namespace std;

static vector<vector<int>> v;
static vector<int> parent;
static vector<int> depth;

static int lca(int a, int b) {
    int depthA = depth[a];
    int depthB = depth[b];

    while (depthA > depthB) {
        a = parent[a];
        depthA--;
    }

    while (depthB > depthA) {
        b = parent[b];
        depthB--;
    }

    while (a != b) {
        a = parent[a];
        b = parent[b];
    }

    return a;
}

static void dfs(int node, int val, int p) {
    depth[node] = val;
    parent[node] = p;

    for (int next : v[node]) {
        if (next != p) {
            dfs(next, val + 1, node);
        }
    }
}

void boj_11437() {
    int n;
    cin >> n;

    v = vector<vector<int>>(n + 1, vector<int>());
    parent = vector<int>(n + 1, 0);
    depth = vector<int>(n + 1, 0);

    for (int i = 1; i < n; i++) {
        int a, b;
        cin >> a >> b;

        v[a].push_back(b);
        v[b].push_back(a);
    }

    dfs(1, 1, 0);

    int m;
    cin >> m;

    for (int i = 0; i < m; i++) {
        int a, b;
        cin >> a >> b;

        cout << lca(a, b) << "\n";
    }
}

File: v1/Kotlin/src/main/kotlin/boj/intermediate/boj_11437.kt
package boj.intermediate

private var n = 0
private var m = 0
private lateinit var tree: Array<ArrayList<Int>>
private lateinit var depth: IntArray
private lateinit var parents: IntArray
private lateinit var visited: BooleanArray

private fun lca(a: Int, b: Int): Int {
    var na = a
    var nb = b

    while (depth[na] != depth[nb]) {
        if (depth[na] > depth[nb]) na = parents[na]
        else nb = parents[nb]
    }

    while (na != nb) {
        na = parents[na]
        nb = parents[nb]
    }

    return na
}

private fun dfs(node: Int, dth: Int) {
    visited[node] = true
    depth[node] = dth

    for (next in tree[node]) {
        if (!visited[next]) {
            parents[next] = node
            dfs(next, dth + 1)
        }
    }
}

fun boj_11437() = with(System.`in`.bufferedReader()) {
    val sb = StringBuilder()

    n = readLine().toInt()
    tree = Array(n + 1) { ArrayList() }
    depth = IntArray(n + 1) { 0 }
    parents = IntArray(n + 1) { 0 }
    visited = BooleanArray(n + 1) { false }

    repeat(n - 1) {
        val (a, b) = readLine().split(" ").map { it.toInt() }
        tree[a] += b
        tree[b] += a
    }

    dfs(1, 0)
    m = readLine().toInt()

    repeat(m) {
        val (a, b) = readLine().split(" ").map { it.toInt() }
        sb.append(lca(a, b)).append('\n')
    }

    print(sb)
    close()
}